lib/modules/manager/cargo/artifacts.ts

```diff
@@ -144,13 +161,32 @@ async function updateArtifactsImpl(
           config.constraints?.rust,
         );
       } else {
-        // If all dependencies have locked versions then update them precisely.
-        await cargoUpdatePrecise(
-          packageFileName,
-          updatedDeps,
-          config.env ?? {},
-          config.constraints?.rust,
+
+        const anyNonLockfileUpdate = updatedDeps.some(
+          (dep) => dep.updateType !== 'lockfileUpdate',
         );
+        if (anyNonLockfileUpdate) {
+          // Cargo fetch is safer to use for non-lockfile updates, because it
+          // has fewer issues with duplicate dependencies.
+          await cargoFetch(
+            packageFileName,
+            config.env ?? {},
+            config.constraints?.rust,
+          );
```

Comment:
We prefer to have a single exec with all commands, so that if there's a failure we can capture it more easily